PR: https://github.com/odoo/odoo/pull/205542

From: 164b55c324e63c45339088823e3d55dfe2147b61
From: Einar Andersen (eian)
Date: 2025-08-22 10:53:17

Breaking data model changes scores: del:7 + add:1, change matches:
+    current_applicant_skill_ids = fields.One2many(
-    matching_score = fields.Float(
-    skill_id = fields.Many2one(
-    skill_level_id = fields.Many2one(
-    skill_type_id = fields.Many2one(
-    level_progress = fields.Integer(
-    skill_ids = fields.Many2many(comodel_name='hr.skill', string="Expected Skills")

Total Changes: 269

[IMP] hr_recruitment: skills v2

This PR introduces new and improved behaviours for managing skills on
models related to the recruitment application. These changes include:
* The ability to add expected skills to a job position.
* A new widget for managing the expected skills on a job position
* Improved dialog for creating skills both on an applicant and job position.
* Improved form view for job positions
* Improved form view for applicants
* Added a new field, score, on education degrees for better job position
  match calculation
* A new field to see how well an applicant matches a job position.
* The ability to browse matching positions from an applicant.
* A new cron to create activities for employees that are missing skills
  that are expected on a job position.
* A new report that allows you to explore how well employees match a job
  position in respect to their skills.
* Other small details like improved demo data, improved views, etc.

task-4680302

closes odoo/odoo#205542

Related: odoo/enterprise#83225
Related: odoo/upgrade#7858
Signed-off-by: Bertrand Dossogne (bedo) <bedo@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_recruitment_skills/models/hr_applicant.py
+++ b/addons/hr_recruitment_skills/models/hr_applicant.py
@@ -11,6 +11,12 @@ class HrApplicant(models.Model):
     applicant_skill_ids = fields.One2many(
         "hr.applicant.skill", "applicant_id", string="Skills", copy=True
     )
+    current_applicant_skill_ids = fields.One2many(
+        comodel_name="hr.applicant.skill",
+        inverse_name="applicant_id",
+        compute="_compute_current_applicant_skill_ids",
+        readonly=False,
+    )
     skill_ids = fields.Many2many("hr.skill", compute="_compute_skill_ids", store=True)
     matching_skill_ids = fields.Many2many(
         comodel_name="hr.skill",
@@ -22,33 +28,52 @@ class HrApplicant(models.Model):
         string="Missing Skills",
         compute="_compute_matching_skill_ids",
     )
-    matching_score = fields.Float(
-        string="Matching Score(%)", compute="_compute_matching_skill_ids"
-    )
+    matching_score = fields.Integer(string="Matching Score", compute="_compute_matching_skill_ids")
+
+    @api.depends("applicant_skill_ids")
+    def _compute_current_applicant_skill_ids(self):
+        current_applicant_skill_by_applicant = self.applicant_skill_ids._get_current_skills_by_applicant()
+        for applicant in self:
+            applicant.current_applicant_skill_ids = current_applicant_skill_by_applicant[applicant.id]
 
     @api.depends("applicant_skill_ids.skill_id")
     def _compute_skill_ids(self):
         for applicant in self:
             applicant.skill_ids = applicant.applicant_skill_ids.skill_id
 
-    @api.depends_context("active_id")
-    @api.depends("skill_ids")
+    @api.depends_context("matching_job_id")
+    @api.depends("current_applicant_skill_ids", "type_id", "job_id", "job_id.job_skill_ids", "job_id.expected_degree")
     def _compute_matching_skill_ids(self):
-        job_id = self.env.context.get("active_id")
-        if not job_id:
-            self.matching_skill_ids = False
-            self.missing_skill_ids = False
-            self.matching_score = 0
-        else:
-            for applicant in self:
-                job_skills = self.env["hr.job"].browse(job_id).skill_ids
-                applicant.matching_skill_ids = job_skills & applicant.skill_ids
-                applicant.missing_skill_ids = job_skills - applicant.skill_ids
-                applicant.matching_score = (
-                    (len(applicant.matching_skill_ids) / len(job_skills)) * 100
-                    if job_skills
-                    else 0
-                )
+        matching_job_id = self.env.context.get("matching_job_id")
+        matching_job = self.env["hr.job"].browse(matching_job_id)
+        for applicant in self:
+            job = matching_job or applicant.job_id
+            if not job or not (job.job_skill_ids or job.expected_degree):
+                applicant.matching_skill_ids = False
+                applicant.missing_skill_ids = False
+                applicant.matching_score = False
+                continue
+            job_skills = job.job_skill_ids
+            job_degree = job.expected_degree.score * 100
+            job_total = sum(job_skills.mapped("level_progress")) + job_degree
+            job_skill_map = {js.skill_id: js.level_progress for js in job_skills}
+
+            matching_applicant_skills = applicant.current_applicant_skill_ids.filtered(
+                lambda a: a.skill_id in job_skill_map,
+            )
+            applicant_degree = applicant.type_id.score * 100 if job_degree > 1 else 0
+            applicant_total = (
+                sum(min(skill.level_progress, job_skill_map[skill.skill_id] * 2) for skill in matching_applicant_skills)
+                + applicant_degree
+            )
+
+            matching_skill_ids = matching_applicant_skills.mapped("skill_id")
+            missing_skill_ids = job_skills.mapped("skill_id") - matching_applicant_skills.mapped("skill_id")
+            matching_score = round(applicant_total / job_total * 100)
+
+            applicant.matching_skill_ids = matching_skill_ids
+            applicant.missing_skill_ids = missing_skill_ids
+            applicant.matching_score = matching_score
 
     def _get_employee_create_vals(self):
         vals = super()._get_employee_create_vals()
@@ -75,50 +100,42 @@ class HrApplicant(models.Model):
         to pass it directly (e.g., applicant.pool_applicant_id.write(vals)) won't yield results so we must
         update each tuple to have the correct command and record ID for the talent pool applicant
 
-        Returns:
-            returns a list of create/update/delete commands with skill_ids relevant to the pool_applicant
+        :param vals: list of CREATE, WRITE or UNLINK commands with skill_ids relevant to the applicant
+        :return: list of CREATE, WRITE or UNLINK commands with skill_ids relevant to the pool_applicant
         """
         applicant_skills = {a.id: a.skill_id.id for a in self.applicant_skill_ids}
         applicant_skills_type = {a.id: a.skill_type_id.id for a in self.applicant_skill_ids}
         talent_skills = {a.skill_id.id: a.id for a in self.pool_applicant_id.applicant_skill_ids}
-        translated_skills = []
-        for skill in vals["applicant_skill_ids"]:
-            command = skill[0]
-            record_id = skill[1]
-            if command == 0:
-                values = skill[2]
-                if values["skill_id"] in talent_skills:
-                    translated_skill = Command.update(
-                        talent_skills[values["skill_id"]],
-                        {"skill_level_id": values["skill_level_id"]},
-                    )
-                    translated_skills.append(translated_skill)
-                else:
-                    translated_skills.append(skill)
-            elif command == 1:
-                values = skill[2]
+        mapped_commands = []
+        for command in vals.get("applicant_skill_ids"):
+            command_number = command[0]
+            record_id = command[1]
+            if command_number == Command.UPDATE:
+                values = command[2]
                 if applicant_skills[record_id] in talent_skills:
-                    translated_skill = Command.update(talent_skills[applicant_skills[record_id]], values)
-                    translated_skills.append(translated_skill)
+                    mapped_command = Command.update(talent_skills[applicant_skills[record_id]], values)
+                    mapped_commands.append(mapped_command)
                 else:
-                    translated_skill = Command.create(
+                    mapped_command = Command.create(
                         {
                             "skill_id": applicant_skills[record_id],
                             "skill_type_id": applicant_skills_type[record_id],
                             "skill_level_id": values["skill_level_id"],
-                        }
+                        },
                     )
-                    translated_skills.append(translated_skill)
-            elif command == 2:
+                    mapped_commands.append(mapped_command)
+            elif command_number == Command.DELETE:
                 if applicant_skills[record_id] in talent_skills:
-                    translated_skill = Command.delete(talent_skills[applicant_skills[record_id]])
-                    translated_skills.append(translated_skill)
-        return translated_skills
+                    mapped_command = Command.delete(talent_skills[applicant_skills[record_id]])
+                    mapped_commands.append(mapped_command)
+            else:
+                mapped_commands.append(command)
+        return mapped_commands
 
     def action_add_to_job(self):
         self.with_context(just_moved=True).write(
             {
-                "job_id": self.env["hr.job"].browse(self.env.context.get("active_id")).id,
+                "job_id": self.env["hr.job"].browse(self.env.context.get("matching_job_id")).id,
                 "stage_id": self.env.ref("hr_recruitment.stage_job0").id,
             }
         )
@@ -126,16 +143,25 @@ class HrApplicant(models.Model):
         action["context"] = literal_eval(action["context"].replace("active_id", str(self.job_id.id)))
         return action
 
+    @api.model_create_multi
+    def create(self, vals_list):
+        if not self:
+            # This is required for the talent pool mechanism to work. Duplicating an hr.applicant record without this
+            # check will cause the skills to not be duplicated or disappear randomly.
+            for vals in vals_list:
+                skills = vals.pop("current_applicant_skill_ids", []) + vals.get("applicant_skill_ids", [])
+                vals["applicant_skill_ids"] = self.env["hr.applicant.skill"]._get_transformed_commands(skills, self)
+        return super().create(vals_list)
+
     def write(self, vals):
-        if (
-            "applicant_skill_ids" in vals
-            and self.pool_applicant_id
-            and (not self.is_pool_applicant)
-        ):
+        if "current_applicant_skill_ids" in vals or "applicant_skill_ids" in vals:
+            skills = vals.pop("current_applicant_skill_ids", []) + vals.get("applicant_skill_ids", [])
+            original_vals = vals.copy()
+            original_vals["applicant_skill_ids"] = skills
+            vals["applicant_skill_ids"] = self.env["hr.applicant.skill"]._get_transformed_commands(skills, self)
             for applicant in self:
-                translated_skills = applicant._map_applicant_skill_ids_to_talent_skill_ids(vals)
-                applicant.pool_applicant_id.write(
-                    {"applicant_skill_ids": translated_skills}
-                )
-        res = super().write(vals)
-        return res
+                # Modify the skill values for the talent if it exists
+                if applicant.pool_applicant_id and (not applicant.is_pool_applicant):
+                    mapped_skills = applicant._map_applicant_skill_ids_to_talent_skill_ids(original_vals)
+                    applicant.pool_applicant_id.write({"applicant_skill_ids": mapped_skills})
+        return super().write(vals)

--- a/addons/hr_recruitment_skills/models/hr_applicant_skill.py
+++ b/addons/hr_recruitment_skills/models/hr_applicant_skill.py
@@ -1,68 +1,36 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
+from collections import defaultdict
 
-from odoo import api, fields, models, _
-from odoo.exceptions import ValidationError
+from odoo import fields, models
 
 
 class HrApplicantSkill(models.Model):
     _name = 'hr.applicant.skill'
+    _inherit = "hr.individual.skill.mixin"
     _description = "Skill level for an applicant"
     _rec_name = 'skill_id'
-    _order = "skill_level_id"
+    _order = "skill_type_id, skill_level_id desc"
 
     applicant_id = fields.Many2one(
-        comodel_name='hr.applicant',
+        comodel_name="hr.applicant",
         required=True,
         index=True,
-        ondelete='cascade')
-    skill_id = fields.Many2one(
-        comodel_name='hr.skill',
-        compute='_compute_skill_id',
-        store=True,
-        domain="[('skill_type_id', '=', skill_type_id)]",
-        readonly=False,
-        required=True)
-    skill_level_id = fields.Many2one(
-        comodel_name='hr.skill.level',
-        compute='_compute_skill_level_id',
-        domain="[('skill_type_id', '=', skill_type_id)]",
-        store=True,
-        readonly=False,
-        required=True)
-    skill_type_id = fields.Many2one(
-        comodel_name='hr.skill.type',
-        required=True)
-    level_progress = fields.Integer(
-        related='skill_level_id.level_progress')
-
-    _unique_skill = models.Constraint(
-        'unique (applicant_id, skill_id)',
-        'Two levels for the same skill is not allowed',
+        ondelete="cascade",
     )
 
-    @api.constrains('skill_id', 'skill_type_id')
-    def _check_skill_type(self):
-        for applicant_skill in self:
-            if applicant_skill.skill_id not in applicant_skill.skill_type_id.skill_ids:
-                raise ValidationError(_("The skill %(name)s and skill type %(type)s doesn't match", name=applicant_skill.skill_id.name, type=applicant_skill.skill_type_id.name))
-
-    @api.constrains('skill_type_id', 'skill_level_id')
-    def _check_skill_level(self):
-        for applicant_skill in self:
-            if applicant_skill.skill_level_id not in applicant_skill.skill_type_id.skill_level_ids:
-                raise ValidationError(_("The skill level %(level)s is not valid for skill type: %(type)s", level=applicant_skill.skill_level_id.name, type=applicant_skill.skill_type_id.name))
-
-    @api.depends('skill_type_id')
-    def _compute_skill_id(self):
-        for applicant_skill in self:
-            if applicant_skill.skill_id.skill_type_id != applicant_skill.skill_type_id:
-                applicant_skill.skill_id = False
-
-    @api.depends('skill_id')
-    def _compute_skill_level_id(self):
-        for applicant_skill in self:
-            if not applicant_skill.skill_id:
-                applicant_skill.skill_level_id = False
-            else:
-                skill_levels = applicant_skill.skill_type_id.skill_level_ids
-                applicant_skill.skill_level_id = skill_levels.filtered('default_level') or skill_levels[0] if skill_levels else False
+    def _linked_field_name(self):
+        return "applicant_id"
+
+    def _get_current_skills_by_applicant(self):
+        applicant_skill_grouped = self.grouped(lambda a_s: (a_s.applicant_id, a_s.skill_id))
+        result_dict = defaultdict(lambda: self.env["hr.applicant.skill"])
+        for (applicant, skill), applicant_skills in applicant_skill_grouped.items():
+            filtered_applicant_skills = applicant_skills.filtered(
+                lambda a_s: not a_s.valid_to or a_s.valid_to >= fields.Date.today(),
+            )
+            if skill.skill_type_id.is_certification and not filtered_applicant_skills:
+                most_recent_certification = max(applicant_skills, key=lambda a_s: a_s.valid_to)
+                result_dict[applicant.id] += most_recent_certification
+                continue
+            result_dict[applicant.id] += filtered_applicant_skills
+        return result_dict

--- a/addons/hr_recruitment_skills/models/hr_job.py
+++ b/addons/hr_recruitment_skills/models/hr_job.py
@@ -1,23 +1,56 @@
-from markupsafe import Markup
 from ast import literal_eval
 
-from odoo import fields, models, _
+from markupsafe import Markup
+
+from odoo import api, fields, models
 
 
 class HrJob(models.Model):
     _inherit = "hr.job"
 
-    skill_ids = fields.Many2many(comodel_name='hr.skill', string="Expected Skills")
+    applicant_matching_score = fields.Float(string="Matching Score(%)", compute="_compute_applicant_matching_score")
+
+    @api.depends_context("active_applicant_id")
+    def _compute_applicant_matching_score(self):
+        active_applicant_id = self.env.context.get("active_applicant_id")
+        if not active_applicant_id:
+            for job in self:
+                job.applicant_matching_score = False
+            return
+
+        applicant = self.env["hr.applicant"].browse(active_applicant_id)
+        for job in self:
+            if not job.job_skill_ids:
+                job.applicant_matching_score = False
+                continue
+            job_skills = job.job_skill_ids
+            job_degree = job.expected_degree.score * 100
+            job_total = sum(job.job_skill_ids.mapped("level_progress")) + job_degree
+            job_skill_map = {js.skill_id.id: js.level_progress for js in job_skills}
+
+            matching_applicant_skills = applicant.current_applicant_skill_ids.filtered(
+                lambda a: a.skill_id.id in job_skill_map,
+            )
+            applicant_degree = applicant.type_id.score * 100 if job_degree > 1 else 0
+            applicant_total = (
+                sum(
+                    min(skill.level_progress, job_skill_map[skill.skill_id.id] * 2)
+                    for skill in matching_applicant_skills
+                )
+                + applicant_degree
+            )
+
+            job.applicant_matching_score = applicant_total / job_total * 100
 
     def action_search_matching_applicants(self):
         self.ensure_one()
-        help_message_1 = _("No Matching Applicants")
-        help_message_2 = _("We do not have any applicants who meet the skill requirements for this job position in the database at the moment.")
+        help_message_1 = self.env._("No Matching Applicants")
+        help_message_2 = self.env._("We do not have any applicants who meet the skill requirements for this job position in the database at the moment.")
         action = self.env['ir.actions.actions']._for_xml_id('hr_recruitment.crm_case_categ0_act_job')
         context = literal_eval(action['context'])
-        context['active_id'] = self.id
+        context['matching_job_id'] = self.id
         action.update({
-            'name': _("Matching Applicants"),
+            'name': self.env._("Matching Applicants"),
             'views': [
                 (self.env.ref('hr_recruitment_skills.crm_case_tree_view_inherit_hr_recruitment_skills').id, 'list'),
                 (False, 'form'),
@@ -25,7 +58,7 @@ class HrJob(models.Model):
             'context': context,
             'domain': [
                 ('job_id', '!=', self.id),
-                ('skill_ids', 'in', self.skill_ids.ids)
+                ('skill_ids', 'in', self.job_skill_ids.skill_id.ids),
             ],
             'help': Markup("<p class='o_view_nocontent_empty_folder'>%s</p><p>%s</p>") % (help_message_1, help_message_2),
         })
